feat(agent/devices): Write the device_ip and the agent_name in the results endpoint

rene-oromtz opened the pull request:
## Description
This PR is meant to add additional information to the results endpoint. Information like `device_ip` and `agent_name` are particular relevant during the Reserve stage. 

If the user is not constantly polling or if the local console where the reservation was executed is closed, the device IP is not able to be located from the current endpoints. 

The proposed approach is a generalization of the current communication method between agent and device made by the multi device agents: 
1. DeviceConnector class will now have a constructor (`__init__`) that expects the configuration from the `default.yaml` file
2. Upon initialization, the device connector will write the device information to `device-info.json` in the run directory
3. The agent will look for the information on this file.
4. If the data can be found, it will send it to the results endpoint.  
5. For user, this information can be retrieved with current `testflinger-cli results` command

This is a non breaking change as all Devices have a `default.yaml` that is retrieved from the commands arguments. 
<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
-->

## Resolved issues
Resolves [CERTTF-632](https://warthogs.atlassian.net/browse/CERTTF-632)
<!--
- Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
- Make sure that the linked issue titles & descriptions are also up to date.
-->

## Documentation

<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Public documentation is in the repository in the docs/ folder.
  - If there impacts on Canonical processes the relevant documentation should be update outside the repository, confirm that this has happened.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
-->

## Web service API changes

<!--
- Are there new endpoints introduced? Please detail for each of them...
  - The rationale for introducing it
  - The interface
    - the HTTP verb
    - the URL structure
    - Versioning
      - Is it expected to be long time stable? It should be versioned (e.g. `.../v2/...`)
      - ... or it expected to evolve, perhaps expected to be a system internal need or something we are expecting to iterate on still? It should be marked unstable (e.g. `.../unstable/...`)
    - the possible request body
    - the response bodies and status code(s) for success and possible failure case(s)
  - Authorization
    - What credentials are required to access the resource?
    - What automated test scenarios are included for authorization failures?
- Are there changed database queries? (... which could cause performance regressions)
- Are there required configuration changes?
- Are there DB migrations included?
- ... or other things you'd like to be aware of at deploy time?
-->

## Tests
This was executed in the staging instance.
<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
-->

Submit job:
```
testflinger-cli --server=https://testflinger-staging.canonical.com submit ~/Desktop/staging.yaml 
Job submitted successfully!
job_id: 4f336dfe-cb1e-4038-8273-6edc6a087fcf
```

cURL from the result endpoint:
```
curl https://testflinger-staging.canonical.com/v1/result/4f336dfe-cb1e-4038-8273-6edc6a087fcf
{"device_info": {"device_ip": "10.2xx.x.xx", "agent_name": "audino"}, "job_state": "reserve"}
```

Testflinger CLI also already includes a results command where this information can be retrieved:
```
testflinger-cli --server=https://testflinger-staging.canonical.com results 4f336dfe-cb1e-4038-8273-6edc6a087fcf
{
    "device_info": {
        "agent_name": "audino",
        "device_ip": "10.2xx.x.xx"
    },
    "job_state": "reserve"
}
```
Note: IP was hidden but it matches the IP available from the charm config.

Only tested on MAAS but writing to the device_info.json should also be possible in all other device types as the only thing needed during provisioning is the config file `default.yaml` where the device_ip and agent_name can be found

[CERTTF-632]: https://warthogs.atlassian.net/browse/CERTTF-632?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

val500 reviewed (changes requested):
I think this looks good, but a unit test in the agent would be nice. You can create a fake device-info as if it was from the device-connector and read from that.

rene-oromtz commented:
Thanks for the review and comments! 
I ended up adding a couple, one for general read of device-info.json and another to test if during reserve phase, the data can be send to the results endpoint.

boukeas commented:
I may be missing quite a bit because I haven't looked into this code in a while but please answer a question for me: why does the device connector need to wait until the end of the provisioning phase to dump the information to that file? It's name and IP should be available right at the time when the device connector starts. Similar question for the agent: as soon as it picks up a job it could submit the information (agent name and device IP) to the appropriate endpoint. Why does it need to wait until it starts the test phase?

And if this concern is valid, we don't have to change every device connector, we only have to change the base one (perhaps even in its `__init__`?)

rene-oromtz commented:
Hi @boukeas 
After our discussion, I ended up giving  it a try and reformatting the code to not depend on provision stage, the write_device_info() function is now part of the `__init__` file and called upon cmd execution. 

As we expected, with this approach the device connector will now write its ip/agent name upon execution of each phase. 

Tested again on staging:
```
testflinger-cli results bffbf633-3c86-4d88-9378-2a0ed11d394c
{
    "device_info": {
        "agent_name": "audino",
        "device_ip": "10.2xx.x.xx"
    },
    "job_state": "reserve"
}
```

boukeas commented on device-connectors/src/testflinger_device_connectors/__init__.py:
```diff
@@ -451,3 +451,21 @@ def _run_test_cmds_str(cmds, config=None, env=None):
+
+    :param config: Dictionary with the key/values from the config file.
+    """
+    try:
+        device_ip = config["device_ip"]
+        agent_name = config["agent_name"]
+        device_info = {
+            "device_info": {"device_ip": device_ip, "agent_name": agent_name}
+        }
+    except (TypeError, KeyError):
+        # If config file not provided or key unable to be located
+        device_info = {}
```
Some questions/concerns:
- I guess a `TypeError` can be raised by these commands only if `config` is not a dict. But then the question is if we should be catching that error. By the way, I don't think that could happen if `write_device_info` is a method of DefaultDevice because then it has access to the configuration file.
- If _one_ of device IP or agent name is missing then _none_ are included in `device_info`. My suggestion:
```
device_info = {
    "device_info": {
        field: value
        for field in ("device_ip", "agent_name")
        if (value := config.get(field))
    }
}
```
This is not exactly equivalent, since `device_info` will always have a `device_info` field, but it might be empty (which is not the case in your original code).

boukeas reviewed (changes requested):
Please let me know if these suggestions make sense to you.

rene-oromtz commented on agent/src/testflinger_agent/job.py:
```diff
@@ -250,6 +258,24 @@ def banner(self, line):
+        """
+        if rundir:
+            try:
+                device_info_file = os.path.join(rundir, "device-info.json")
```
No particular reason, I was using the `os.path` because it was already in use by other portions of the code. 
I included the `pathlib.Path ` as this might be preferred.

rene-oromtz commented:
@boukeas 
Based on our latest discussion, current approach is to create a method that gets called upon initialization of the `DefaultDevice` class. 

The minimum modification approach is to initialize the class with the config file that is available for all devices/agents. No modifications on the DeviceConnectors for now (only on Unit Tests). As agreed, we can create a follow up PR to allow all DeviceConnector to access the new attribute `config` to avoid opening this file on every module. 

Please let me know what you think and if there is anything else that needs modification.

rene-oromtz commented:
@boukeas 

Thanks for the suggestions, I added the requested changes. 

> Please include tests from a "regular" job (e.g. provisioning and a trivial test phase) and a reserve job. Is it also possible to test a multi-device job?

I added a parametrized UT that validates if all phases completes successfully. For the multi-device agent, there is already a UT that kind of do this: `test_wait_for_completion`. This doesn't run the phase directly but call the `wait_for_completion()` which validates if the parent_id is already completed. 

> Check if we need an update to the docs as well.

Thanks for the advise! I added in a section where it was mentioning the use of the `results` command. Please let me know if you agree of the phrasing.
